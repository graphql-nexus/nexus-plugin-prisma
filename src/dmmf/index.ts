export { get } from './importer'
export { transform } from './transformer'
export { DMMFClass as DMMF } from './DMMFClass'
export { ExternalDMMF as External } from './types'
