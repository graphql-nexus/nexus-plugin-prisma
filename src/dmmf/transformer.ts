import { DMMF, ExternalDMMF } from './types'

export function transform(document: DMMF.Document): ExternalDMMF.Document {
  return {
    datamodel: transformDatamodel(document.datamodel),
    mappings: document.mappings as ExternalDMMF.Mapping[],
    schema: transformSchema(document.schema),
  }
}

function transformDatamodel(datamodel: DMMF.Datamodel): ExternalDMMF.Datamodel {
  return {
    enums: datamodel.enums,
    models: datamodel.models.map(model => ({
      ...model,
      fields: model.fields.map(field => ({
        ...field,
        kind: field.kind === 'object' ? 'relation' : field.kind,
      })),
    })),
  }
}

function transformSchema(schema: DMMF.Schema): ExternalDMMF.Schema {
  return {
    enums: schema.enums,
    inputTypes: schema.inputTypes.map(transformInputType),
    outputTypes: schema.outputTypes.map(o => ({
      ...o,
      fields: o.fields.map(f => ({
        ...f,
        args: f.args.map(transformArg),
        outputType: {
          ...f.outputType,
          type: getReturnTypeName(f.outputType.type),
        },
      })),
    })),
  }
}

function transformArg(arg: DMMF.SchemaArg): ExternalDMMF.SchemaArg {
  // FIXME: *Enum*Filter are currently empty
  let inputType = arg.inputType.some(a => a.kind === 'enum')
    ? arg.inputType[0]
    : arg.inputType.find(a => a.kind === 'object')!

  if (!inputType) {
    inputType = arg.inputType[0]
  }

  return {
    name: arg.name,
    inputType: {
      ...inputType,
      type: getReturnTypeName(inputType.type),
    },
    isRelationFilter: undefined,
  }
}

function transformInputType(inputType: DMMF.InputType): ExternalDMMF.InputType {
  return {
    ...inputType,
    fields: inputType.fields.map(transformArg),
  }
}

function getReturnTypeName(type: any) {
  if (typeof type === 'string') {
    return type
  }

  return type.name
}
