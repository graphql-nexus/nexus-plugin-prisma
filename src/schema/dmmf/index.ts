export * from './utils'
export * from './transformer'
export * from './DmmfDocument'
export * from './DmmfTypes'
