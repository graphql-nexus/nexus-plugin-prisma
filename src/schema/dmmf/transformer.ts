import { DMMF } from '@prisma/client/runtime'
import { paginationStrategies, PaginationStrategy } from '../pagination'
import { GlobalComputedInputs, GlobalMutationResolverParams, LocalComputedInputs } from '../utils'
import { DmmfDocument } from './DmmfDocument'
import { DmmfTypes } from './DmmfTypes'
import { getPrismaClientDmmf } from './utils'

export type TransformOptions = {
  globallyComputedInputs?: GlobalComputedInputs
  paginationStrategy?: PaginationStrategy
}

export const getTransformedDmmf = (
  prismaClientPackagePath: string,
  options?: TransformOptions
): DmmfDocument => new DmmfDocument(transform(getPrismaClientDmmf(prismaClientPackagePath), options))

const addDefaultOptions = (givenOptions?: TransformOptions): Required<TransformOptions> => ({
  globallyComputedInputs: {},
  paginationStrategy: paginationStrategies.relay,
  ...givenOptions,
})

export function transform(document: DMMF.Document, options?: TransformOptions): DmmfTypes.Document {
  return {
    datamodel: transformDatamodel(document.datamodel),
    mappings: document.mappings as DmmfTypes.Mapping[],
    schema: transformSchema(document.schema, addDefaultOptions(options)),
  }
}

function transformDatamodel(datamodel: DMMF.Datamodel): DmmfTypes.Datamodel {
  return {
    enums: datamodel.enums,
    models: datamodel.models.map((model) => ({
      ...model,
      fields: model.fields.map((field) => ({
        ...field,
        kind: field.kind === 'object' ? 'relation' : field.kind,
      })),
    })) as any, // TODO: Remove this once @prisma/client/runtime:DMMF contains the `uniqueFields` typed
  }
}

const paginationArgNames = ['cursor', 'take', 'skip']

function transformSchema(
  schema: DMMF.Schema,
  { globallyComputedInputs, paginationStrategy }: Required<TransformOptions>
): DmmfTypes.Schema {
  return {
    enums: schema.enums,
    inputTypes: schema.inputTypes.map((_) => transformInputType(_, globallyComputedInputs)),
    outputTypes: schema.outputTypes.map((o) => {
      return {
        ...o,
        fields: o.fields.map((f) => {
          let args = f.args.map(transformArg)
          const argNames = args.map((a) => a.name)

          // If this field has pagination
          if (paginationArgNames.every((paginationArgName) => argNames.includes(paginationArgName))) {
            args = paginationStrategy.transformDmmfArgs({
              args,
              paginationArgNames,
              field: f,
            })
          }

          return {
            ...f,
            args,
            outputType: {
              ...f.outputType,
              type: getReturnTypeName(f.outputType.type) as any,
            },
          }
        }),
      }
    }),
  }
}

/**
 * Conversion from a Prisma Client arg type to a GraphQL arg type using
 * heuristics. A conversion is needed because GraphQL does not
 * support union types on args, but Prisma Client does.
 */
function transformArg(arg: DMMF.SchemaArg): DmmfTypes.SchemaArg {
  // FIXME: *Enum*Filter are currently empty
  let inputType = arg.inputType.some((a) => a.kind === 'enum')
    ? arg.inputType[0]
    : arg.inputType.find((a) => a.kind === 'object')!

  if (!inputType) {
    inputType = arg.inputType[0]
  }

  return {
    name: arg.name,
    inputType: {
      ...inputType,
      type: getReturnTypeName(inputType.type),
    },
    // FIXME Why?
    isRelationFilter: undefined,
  }
}

type AddComputedInputParams = {
  inputType: DmmfTypes.InputType
  params: GlobalMutationResolverParams
  dmmf: DmmfDocument
  locallyComputedInputs: LocalComputedInputs<any>
}

/** Resolver-level computed inputs aren't recursive so aren't
 *  needed for deep computed inputs.
 */
type AddDeepComputedInputsArgs = Omit<AddComputedInputParams, 'locallyComputedInputs'> & { data: any } // Used to recurse through the input object

/**
 * Recursively looks for inputs that need a value from globallyComputedInputs
 * and populates them
 */
function addGloballyComputedInputs({
  inputType,
  params,
  dmmf,
  data,
}: AddDeepComputedInputsArgs): Record<string, any> {
  if (Array.isArray(data)) {
    return data.map((value) =>
      addGloballyComputedInputs({
        inputType,
        dmmf,
        params,
        data: value,
      })
    )
  }
  // Get values for computedInputs corresponding to keys that exist in inputType
  const computedInputValues = Object.keys(inputType.computedInputs).reduce(
    (values, key) => ({
      ...values,
      [key]: inputType.computedInputs[key](params),
    }),
    {} as Record<string, any>
  )
  // Combine computedInputValues with values provided by the user, recursing to add
  // global computedInputs to nested types
  return Object.keys(data).reduce((deeplyComputedData, fieldName) => {
    const field = inputType.fields.find((_) => _.name === fieldName)!
    const fieldValue =
      field.inputType.kind === 'object'
        ? addGloballyComputedInputs({
            inputType: dmmf.getInputType(field.inputType.type),
            dmmf,
            params,
            data: data[fieldName],
          })
        : data[fieldName]
    return {
      ...deeplyComputedData,
      [fieldName]: fieldValue,
    }
  }, computedInputValues)
}

export function addComputedInputs({
  dmmf,
  inputType,
  locallyComputedInputs,
  params,
}: AddComputedInputParams) {
  return {
    ...params.args,
    data: {
      /**
       * Globally computed inputs are attached to the inputType object
       * as 'computedInputs' by the transformInputType function.
       */
      ...addGloballyComputedInputs({
        inputType,
        dmmf,
        params,
        data: params.args.data,
      }),
      ...Object.entries(locallyComputedInputs).reduce(
        (args, [fieldName, computeFieldValue]) => ({
          ...args,
          [fieldName]: computeFieldValue(params),
        }),
        {} as Record<string, any>
      ),
    },
  }
}

function transformInputType(
  inputType: DMMF.InputType,
  globallyComputedInputs: GlobalComputedInputs
): DmmfTypes.InputType {
  const fieldNames = inputType.fields.map((field) => field.name)
  /**
   * Only global computed inputs are removed during schema transform.
   * Resolver level computed inputs are filtered as part of the
   * projecting process. They are then passed to addComputedInputs
   * at runtime so their values can be inferred alongside the
   * global values.
   */
  const globallyComputedInputsInType = Object.keys(globallyComputedInputs).reduce(
    (args, key) =>
      fieldNames.includes(key) ? Object.assign(args, { [key]: globallyComputedInputs[key] }) : args,
    {} as GlobalComputedInputs
  )
  return {
    ...inputType,
    fields: inputType.fields.filter((field) => !(field.name in globallyComputedInputs)).map(transformArg),
    computedInputs: globallyComputedInputsInType,
  }
}

/**
 * Make the "return type" property type always be a string. In Prisma Client
 * it is allowed to be a nested structured object but we want only the
 * reference-by-name form.
 *
 */
//
// TODO _why_ is the dmmf like this?
//
// FIXME `any` type because this is used by both outputType and inputType
// and there is currently no generic capturing both ideas.
//
export function getReturnTypeName(type: any): string {
  if (typeof type === 'string') {
    return type
  }

  return type.name
}
