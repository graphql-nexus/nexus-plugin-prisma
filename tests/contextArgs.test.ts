import * as Nexus from 'nexus'
import { generateSchemaAndTypesWithoutThrowing, getDmmf } from './__utils'

const resolverTestData = {
  datamodel: `
  model User {
    id  Int @id
    name String
    browser String
  }
`,
  query: Nexus.objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.user()
    },
  }),
  mutation: Nexus.objectType({
    name: 'Mutation',
    definition(t: any) {
      t.crud.createOneUser({
        contextArgs: { browser: (ctx: any) => ctx.browser },
      })
    },
  }),
  user: Nexus.objectType({
    name: 'User',
    definition: (t: any) => {
      t.model.id()
      t.model.name()
      t.model.browser()
    },
  }),
}

const globalTestData = {
  datamodel: `
  model User {
    id  Int @id
    name String
    nested Nested[]
    browser String
  }

  model Nested {
    id Int @id
    browser String
  }
`,
  query: Nexus.objectType({
    name: 'Query',
    definition(t: any) {
      t.crud.user()
    },
  }),
  mutation: Nexus.objectType({
    name: 'Mutation',
    definition(t: any) {
      t.crud.createOneUser()
      t.crud.createOneNested()
    },
  }),
  user: Nexus.objectType({
    name: 'User',
    definition: (t: any) => {
      t.model.id()
      t.model.name()
      t.model.nested()
      t.model.browser()
    },
  }),
  nested: Nexus.objectType({
    name: 'Nested',
    definition: (t: any) => {
      t.model.id()
      t.model.browser()
    },
  }),
}

it('removes resolver-level contextArgs from the corresponding input type', async () => {
  const { datamodel, ...resolvers } = resolverTestData
  const result = await generateSchemaAndTypesWithoutThrowing(
    datamodel,
    Object.values(resolvers),
  )

  expect(result).toMatchSnapshot('resolverContextArgs')
})

import { addContextArgs } from '../src/dmmf/transformer'

it('infers the value of resolver-level contextArgs at runtime', async () => {
  const { datamodel } = resolverTestData
  const dmmf = await getDmmf(datamodel)
  expect(
    addContextArgs({
      baseArgs: { data: { name: 'New User' } },
      ctx: { browser: 'firefox' },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      contextArgs: { browser: (ctx: any) => ctx.browser },
    }),
  ).toStrictEqual({ data: { name: 'New User', browser: 'firefox' } })
})

it('removes global contextArg fields from all input types', async () => {
  const { datamodel, ...resolvers } = globalTestData
  const result = await generateSchemaAndTypesWithoutThrowing(
    datamodel,
    Object.values(resolvers),
    { contextArgs: { browser: (ctx: any) => ctx.browser } },
  )

  expect(result).toMatchSnapshot('globalContextArgs')
})

it('infers the value of global contextArgs at runtime', async () => {
  const { datamodel } = globalTestData
  const dmmf = await getDmmf(datamodel, {
    contextArgs: { browser: (ctx: any) => ctx.browser },
  })
  expect(
    addContextArgs({
      baseArgs: { data: { name: 'New User', nested: { create: {} } } },
      ctx: { browser: 'firefox' },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      contextArgs: {},
    }),
  ).toStrictEqual({
    data: {
      name: 'New User',
      browser: 'firefox',
      nested: { create: { browser: 'firefox' } },
    },
  })
})

it('handles arrays when recursing for contextArgs', async () => {
  const { datamodel } = globalTestData
  const dmmf = await getDmmf(datamodel, {
    contextArgs: { browser: (ctx: any) => ctx.browser },
  })
  expect(
    addContextArgs({
      baseArgs: { data: { name: 'New User', nested: { create: [{}, {}] } } },
      ctx: { browser: 'firefox' },
      inputType: dmmf.getInputType('UserCreateInput'),
      dmmf,
      contextArgs: {},
    }),
  ).toStrictEqual({
    data: {
      name: 'New User',
      browser: 'firefox',
      nested: { create: [{ browser: 'firefox' }, { browser: 'firefox' }] },
    },
  })
})

it('Works on Redo', async () => {
  const dmmf = await getDmmf(
    `model Tag {
  id   Int    @id
  user User
  name String
  @@unique([name, user])
}

model Selector {
  id   Int    @id
  user User
  css  String
}

model Step {
  id       Int      @id
  user     User
  action   String
  selector Selector
  value    String
}

model Test {
  id    Int    @id
  user  User
  name  String
  steps Step[]
  tags  Tag[]
  @@unique([name, user])
}

model User {
  id       Int    @id
  email    String @unique
  password String
  first    String
  last     String
}`,
    { contextArgs: { user: (ctx: any) => ({ connect: { id: ctx.userId } }) } },
  )
  expect(
    addContextArgs({
      baseArgs: {
        data: {
          name: 'Test 1',
          steps: { create: [{ action: 'click', value: 'something' }] },
        },
      },
      ctx: { userId: 1 },
      inputType: dmmf.getInputType('TestCreateInput'),
      dmmf,
      contextArgs: {},
    }),
  ).toStrictEqual({
    data: {
      name: 'Test 1',
      steps: {
        create: [
          { action: 'click', value: 'something', user: { connect: { id: 1 } } },
        ],
      },
      user: { connect: { id: 1 } },
    },
  })
})
