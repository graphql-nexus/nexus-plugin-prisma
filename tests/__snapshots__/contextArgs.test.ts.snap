// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removes global contextArg fields from all input types: globalContextArgs 1`] = `
Object {
  "missingTypes": Object {},
  "schema": "type Mutation {
  createOneUser(data: UserCreateInput!): User!
  createOneNested(data: NestedCreateInput!): Nested!
}

type Nested {
  id: Int!
  createdWithBrowser: String!
  name: String!
}

input NestedCreateInput {
  name: String!
  user: UserCreateOneWithoutUserInput
}

input NestedCreateManyWithoutNestedInput {
  create: [NestedCreateWithoutUserInput!]
  connect: [NestedWhereUniqueInput!]
}

input NestedCreateWithoutUserInput {
  name: String!
}

input NestedWhereUniqueInput {
  id: Int
}

type Query {
  user(where: UserWhereUniqueInput!): User
}

type User {
  id: Int!
  name: String!
  nested(skip: Int, after: Int, before: Int, first: Int, last: Int): [Nested!]!
  createdWithBrowser: String!
}

input UserCreateInput {
  name: String!
  nested: NestedCreateManyWithoutNestedInput
}

input UserCreateOneWithoutUserInput {
  create: UserCreateWithoutNestedInput
  connect: UserWhereUniqueInput
}

input UserCreateWithoutNestedInput {
  name: String!
}

input UserWhereUniqueInput {
  id: Int
}
",
  "typegen": "import * as photon from '@generated/photon';
import { core } from 'nexus';
// Types helpers
  type IsModelNameExistsInGraphQLTypes<
  ReturnType extends any
> = ReturnType extends core.GetGen<'objectNames'> ? true : false;

type NexusPrismaScalarOpts = {
  alias?: string;
};

type Pagination = {
  first?: boolean;
  last?: boolean;
  before?: boolean;
  after?: boolean;
  skip?: boolean;
};

type RootObjectTypes = Pick<
  core.GetGen<'rootTypes'>,
  core.GetGen<'objectNames'>
>;

/**
 * Determine if \`B\` is a subset (or equivalent to) of \`A\`.
*/
type IsSubset<A, B> = keyof A extends never
  ? false
  : B extends A
  ? true
  : false;

type OmitByValue<T, ValueType> = Pick<
  T,
  { [Key in keyof T]: T[Key] extends ValueType ? never : Key }[keyof T]
>;

type GetSubsetTypes<ModelName extends any> = keyof OmitByValue<
  {
    [P in keyof RootObjectTypes]: ModelName extends keyof ModelTypes
      ? IsSubset<RootObjectTypes[P], ModelTypes[ModelName]> extends true
        ? RootObjectTypes[P]
        : never
      : never;
  },
  never
>;

type SubsetTypes<ModelName extends any> = GetSubsetTypes<
  ModelName
> extends never
  ? \`ERROR: No subset types are available. Please make sure that one of your GraphQL type is a subset of your t.model('<ModelName>')\`
  : GetSubsetTypes<ModelName>;

type DynamicRequiredType<ReturnType extends any> = IsModelNameExistsInGraphQLTypes<
  ReturnType
> extends true
  ? { type?: SubsetTypes<ReturnType> }
  : { type: SubsetTypes<ReturnType> };

type GetNexusPrismaInput<
  ModelName extends any,
  MethodName extends any,
  InputName extends 'filtering' | 'ordering'
> = ModelName extends keyof NexusPrismaInputs
  ? MethodName extends keyof NexusPrismaInputs[ModelName]
    ? NexusPrismaInputs[ModelName][MethodName][InputName]
    : never
  : never;

type ContextArgs = Record<string, (ctx: any) => any>;

type NexusPrismaRelationOpts<
  ModelName extends any,
  MethodName extends any,
  ReturnType extends any
> = GetNexusPrismaInput<
  // If GetNexusPrismaInput returns never, it means there are no filtering/ordering args for it. So just use \`alias\` and \`type\`
  ModelName,
  MethodName,
  'filtering'
> extends never
  ? {
      alias?: string;
      contextArgs?: ContextArgs;
    } & DynamicRequiredType<ReturnType>
  : {
      alias?: string;
      contextArgs?: ContextArgs;
      filtering?:
        | boolean
        | Partial<
            Record<
              GetNexusPrismaInput<ModelName, MethodName, 'filtering'>,
              boolean
            >
          >;
      ordering?:
        | boolean
        | Partial<
            Record<
              GetNexusPrismaInput<ModelName, MethodName, 'ordering'>,
              boolean
            >
          >;
      pagination?: boolean | Pagination;
    } & DynamicRequiredType<ReturnType>;

type IsScalar<TypeName extends any> = TypeName extends core.GetGen<'scalarNames'>
  ? true
  : false;

type IsObject<Name extends any> = Name extends core.GetGen<'objectNames'>
  ? true
  : false

type IsEnum<Name extends any> = Name extends core.GetGen<'enumNames'>
  ? true
  : false

type IsInputObject<Name extends any> = Name extends core.GetGen<'inputNames'>
  ? true
  : false

/**
 * The kind that a GraphQL type may be.
 */
type Kind = 'Enum' | 'Object' | 'Scalar' | 'InputObject'

/**
 * Helper to safely reference a Kind type. For example instead of the following
 * which would admit a typo:
 *
 * \`\`\`ts
 * type Foo = Bar extends 'scalar' ? ...
 * \`\`\`
 *
 * You can do this which guarantees a correct reference:
 *
 * \`\`\`ts
 * type Foo = Bar extends AKind<'Scalar'> ? ...
 * \`\`\`
 *
 */
type AKind<T extends Kind> = T

type GetKind<Name extends any> = IsEnum<Name> extends true
  ? 'Enum'
  : IsScalar<Name> extends true
  ? 'Scalar'
  : IsObject<Name> extends true
  ? 'Object'
  : IsInputObject<Name> extends true
  ? 'InputObject'
  // FIXME should be \`never\`, but GQL objects named differently
  // than backing type fall into this branch
  : 'Object'

type NexusPrismaFields<ModelName extends keyof NexusPrismaTypes> = {
  [MethodName in keyof NexusPrismaTypes[ModelName]]: NexusPrismaMethod<
    ModelName,
    MethodName,
    GetKind<NexusPrismaTypes[ModelName][MethodName]> // Is the return type a scalar?
  >;
};

type NexusPrismaMethod<
  ModelName extends keyof NexusPrismaTypes,
  MethodName extends keyof NexusPrismaTypes[ModelName],
  ThisKind extends Kind,
  ReturnType extends any = NexusPrismaTypes[ModelName][MethodName]
> =
  ThisKind extends AKind<'Enum'>
  ? () => NexusPrismaFields<ModelName>
  : ThisKind extends AKind<'Scalar'>
  ? (opts?: NexusPrismaScalarOpts) => NexusPrismaFields<ModelName> // Return optional scalar opts
  : IsModelNameExistsInGraphQLTypes<ReturnType> extends true // If model name has a mapped graphql types
  ? (
      opts?: NexusPrismaRelationOpts<ModelName, MethodName, ReturnType>
    ) => NexusPrismaFields<ModelName> // Then make opts optional
  : (
      opts: NexusPrismaRelationOpts<ModelName, MethodName, ReturnType>
    ) => NexusPrismaFields<ModelName>; // Else force use input the related graphql type -> { type: '...' }

type GetNexusPrismaMethod<
  TypeName extends string
> = TypeName extends keyof NexusPrismaMethods
  ? NexusPrismaMethods[TypeName]
  : <CustomTypeName extends keyof ModelTypes>(
      typeName: CustomTypeName
    ) => NexusPrismaMethods[CustomTypeName];

type GetNexusPrisma<
  TypeName extends string,
  ModelOrCrud extends 'model' | 'crud'
> = ModelOrCrud extends 'model'
  ? TypeName extends 'Mutation'
    ? never
    : TypeName extends 'Query'
    ? never
    : GetNexusPrismaMethod<TypeName>
  : ModelOrCrud extends 'crud'
  ? TypeName extends 'Mutation'
    ? GetNexusPrismaMethod<TypeName>
    : TypeName extends 'Query'
    ? GetNexusPrismaMethod<TypeName>
    : never
  : never;
  

// Generated
interface ModelTypes {
  User: photon.User
  Nested: photon.Nested
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'name' | 'nested' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'name'
}
    nesteds: {
  filtering: 'id' | 'name' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name'
}

  },
    User: {
    nested: {
  filtering: 'id' | 'name' | 'AND' | 'OR' | 'NOT' | 'user'
  ordering: 'id' | 'name'
}

  },  Nested: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'
    nested: 'Nested'
    nesteds: 'Nested'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'
    createOneNested: 'Nested'
    updateOneNested: 'Nested'
    updateManyNested: 'BatchPayload'
    deleteOneNested: 'Nested'
    deleteManyNested: 'BatchPayload'
    upsertOneNested: 'Nested'

  },
  User: {
    id: 'Int'
    name: 'String'
    nested: 'Nested'
    createdWithBrowser: 'String'

},  Nested: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'
    user: 'User'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Nested: NexusPrismaFields<'Nested'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;
}
  ",
}
`;

exports[`removes resolver-level contextArgs from the corresponding input type: resolverContextArgs 1`] = `
Object {
  "missingTypes": Object {},
  "schema": "type Mutation {
  createOneUser(data: UserCreateInput!): User!
}

type Query {
  user(where: UserWhereUniqueInput!): User
}

type User {
  id: Int!
  name: String!
  createdWithBrowser: String!
}

input UserCreateInput {
  name: String!
}

input UserWhereUniqueInput {
  id: Int
}
",
  "typegen": "import * as photon from '@generated/photon';
import { core } from 'nexus';
// Types helpers
  type IsModelNameExistsInGraphQLTypes<
  ReturnType extends any
> = ReturnType extends core.GetGen<'objectNames'> ? true : false;

type NexusPrismaScalarOpts = {
  alias?: string;
};

type Pagination = {
  first?: boolean;
  last?: boolean;
  before?: boolean;
  after?: boolean;
  skip?: boolean;
};

type RootObjectTypes = Pick<
  core.GetGen<'rootTypes'>,
  core.GetGen<'objectNames'>
>;

/**
 * Determine if \`B\` is a subset (or equivalent to) of \`A\`.
*/
type IsSubset<A, B> = keyof A extends never
  ? false
  : B extends A
  ? true
  : false;

type OmitByValue<T, ValueType> = Pick<
  T,
  { [Key in keyof T]: T[Key] extends ValueType ? never : Key }[keyof T]
>;

type GetSubsetTypes<ModelName extends any> = keyof OmitByValue<
  {
    [P in keyof RootObjectTypes]: ModelName extends keyof ModelTypes
      ? IsSubset<RootObjectTypes[P], ModelTypes[ModelName]> extends true
        ? RootObjectTypes[P]
        : never
      : never;
  },
  never
>;

type SubsetTypes<ModelName extends any> = GetSubsetTypes<
  ModelName
> extends never
  ? \`ERROR: No subset types are available. Please make sure that one of your GraphQL type is a subset of your t.model('<ModelName>')\`
  : GetSubsetTypes<ModelName>;

type DynamicRequiredType<ReturnType extends any> = IsModelNameExistsInGraphQLTypes<
  ReturnType
> extends true
  ? { type?: SubsetTypes<ReturnType> }
  : { type: SubsetTypes<ReturnType> };

type GetNexusPrismaInput<
  ModelName extends any,
  MethodName extends any,
  InputName extends 'filtering' | 'ordering'
> = ModelName extends keyof NexusPrismaInputs
  ? MethodName extends keyof NexusPrismaInputs[ModelName]
    ? NexusPrismaInputs[ModelName][MethodName][InputName]
    : never
  : never;

type ContextArgs = Record<string, (ctx: any) => any>;

type NexusPrismaRelationOpts<
  ModelName extends any,
  MethodName extends any,
  ReturnType extends any
> = GetNexusPrismaInput<
  // If GetNexusPrismaInput returns never, it means there are no filtering/ordering args for it. So just use \`alias\` and \`type\`
  ModelName,
  MethodName,
  'filtering'
> extends never
  ? {
      alias?: string;
      contextArgs?: ContextArgs;
    } & DynamicRequiredType<ReturnType>
  : {
      alias?: string;
      contextArgs?: ContextArgs;
      filtering?:
        | boolean
        | Partial<
            Record<
              GetNexusPrismaInput<ModelName, MethodName, 'filtering'>,
              boolean
            >
          >;
      ordering?:
        | boolean
        | Partial<
            Record<
              GetNexusPrismaInput<ModelName, MethodName, 'ordering'>,
              boolean
            >
          >;
      pagination?: boolean | Pagination;
    } & DynamicRequiredType<ReturnType>;

type IsScalar<TypeName extends any> = TypeName extends core.GetGen<'scalarNames'>
  ? true
  : false;

type IsObject<Name extends any> = Name extends core.GetGen<'objectNames'>
  ? true
  : false

type IsEnum<Name extends any> = Name extends core.GetGen<'enumNames'>
  ? true
  : false

type IsInputObject<Name extends any> = Name extends core.GetGen<'inputNames'>
  ? true
  : false

/**
 * The kind that a GraphQL type may be.
 */
type Kind = 'Enum' | 'Object' | 'Scalar' | 'InputObject'

/**
 * Helper to safely reference a Kind type. For example instead of the following
 * which would admit a typo:
 *
 * \`\`\`ts
 * type Foo = Bar extends 'scalar' ? ...
 * \`\`\`
 *
 * You can do this which guarantees a correct reference:
 *
 * \`\`\`ts
 * type Foo = Bar extends AKind<'Scalar'> ? ...
 * \`\`\`
 *
 */
type AKind<T extends Kind> = T

type GetKind<Name extends any> = IsEnum<Name> extends true
  ? 'Enum'
  : IsScalar<Name> extends true
  ? 'Scalar'
  : IsObject<Name> extends true
  ? 'Object'
  : IsInputObject<Name> extends true
  ? 'InputObject'
  // FIXME should be \`never\`, but GQL objects named differently
  // than backing type fall into this branch
  : 'Object'

type NexusPrismaFields<ModelName extends keyof NexusPrismaTypes> = {
  [MethodName in keyof NexusPrismaTypes[ModelName]]: NexusPrismaMethod<
    ModelName,
    MethodName,
    GetKind<NexusPrismaTypes[ModelName][MethodName]> // Is the return type a scalar?
  >;
};

type NexusPrismaMethod<
  ModelName extends keyof NexusPrismaTypes,
  MethodName extends keyof NexusPrismaTypes[ModelName],
  ThisKind extends Kind,
  ReturnType extends any = NexusPrismaTypes[ModelName][MethodName]
> =
  ThisKind extends AKind<'Enum'>
  ? () => NexusPrismaFields<ModelName>
  : ThisKind extends AKind<'Scalar'>
  ? (opts?: NexusPrismaScalarOpts) => NexusPrismaFields<ModelName> // Return optional scalar opts
  : IsModelNameExistsInGraphQLTypes<ReturnType> extends true // If model name has a mapped graphql types
  ? (
      opts?: NexusPrismaRelationOpts<ModelName, MethodName, ReturnType>
    ) => NexusPrismaFields<ModelName> // Then make opts optional
  : (
      opts: NexusPrismaRelationOpts<ModelName, MethodName, ReturnType>
    ) => NexusPrismaFields<ModelName>; // Else force use input the related graphql type -> { type: '...' }

type GetNexusPrismaMethod<
  TypeName extends string
> = TypeName extends keyof NexusPrismaMethods
  ? NexusPrismaMethods[TypeName]
  : <CustomTypeName extends keyof ModelTypes>(
      typeName: CustomTypeName
    ) => NexusPrismaMethods[CustomTypeName];

type GetNexusPrisma<
  TypeName extends string,
  ModelOrCrud extends 'model' | 'crud'
> = ModelOrCrud extends 'model'
  ? TypeName extends 'Mutation'
    ? never
    : TypeName extends 'Query'
    ? never
    : GetNexusPrismaMethod<TypeName>
  : ModelOrCrud extends 'crud'
  ? TypeName extends 'Mutation'
    ? GetNexusPrismaMethod<TypeName>
    : TypeName extends 'Query'
    ? GetNexusPrismaMethod<TypeName>
    : never
  : never;
  

// Generated
interface ModelTypes {
  User: photon.User
}
  
interface NexusPrismaInputs {
  Query: {
    users: {
  filtering: 'id' | 'name' | 'createdWithBrowser' | 'AND' | 'OR' | 'NOT'
  ordering: 'id' | 'name' | 'createdWithBrowser'
}

  },
    User: {


  }
}

interface NexusPrismaTypes {
  Query: {
    user: 'User'
    users: 'User'

  },
  Mutation: {
    createOneUser: 'User'
    updateOneUser: 'User'
    updateManyUser: 'BatchPayload'
    deleteOneUser: 'User'
    deleteManyUser: 'BatchPayload'
    upsertOneUser: 'User'

  },
  User: {
    id: 'Int'
    name: 'String'
    createdWithBrowser: 'String'

}
}

interface NexusPrismaMethods {
  User: NexusPrismaFields<'User'>
  Query: NexusPrismaFields<'Query'>
  Mutation: NexusPrismaFields<'Mutation'>
}
  

declare global {
  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = GetNexusPrisma<TypeName, ModelOrCrud>;
}
  ",
}
`;
